docs/assignment3.md

```diff
@@ -1 +1,577 @@
+| **Intended use**       | This allows the client to send complete Instruction objects (in our case across the network) to be executed on the server. When adding new commands to the server, only the server has to be updated as the client will only be sending requests in a generic textual form, not a specialised object that is directly executed on the server. <br>This is also a security measure, as it removes any client-side processing.|
+| **Constraints**        | There is no defined way of sending these commands (yet). The next design pattern will go more in-depth for that. |
+
+| ID                     | DP2 |
+| ---------------------- | ------------------------------------------------------------ |
+| **Design pattern**     | Interpreter Pattern |
+| **Problem**            | With Design Pattern 1 we now have a clear way of packaging requests from a Player to the Game. However, how the requests are structured is another problem. We will need a proper grammar as the server would not know how to interpret a command without this. A grammar or syntax for the language should be specified, preferably easily understandable and extendable. |
+| **Solution**           | A solution, of course, is actually making the standardized language for passing commands to the server, but how to structure it? We decided to use the following syntax: `<action> <item> <preposition> <item>`. The command is a single string, the arguments can contain more strings separated by a space. The server can interpret the first word as a command and then immediately forward the arguments to respective method. If the command is in the wrong syntax, an error should be displayed. |
+| **Intended use**       | When receiving command by the client, they should be in this defined grammar. An example: The command `give wallet guard` is received by the PlayerThread and then forwarded to the Interpreter. The interpreter then selects the correct handling function, in this case `giveItem` and passes on the arguments. In this function the argument syntax is also checked (if applicable), which in this case is `<item> <proposition <item>`. The `give` command in this case would fail as the second argument is not "to". |
+| **Constraints**        | The current code assumes `<action> <item> <preposition> <item>`, which is also like simple commands in natural language. E.g. "give x to y", "drop x", "shout z". If you want to change the command syntax to allow different input, for example more natural language e.g. "I want to give the wallet in my backpack to the guard that is standing at the entrance", this would not be easy. Natural language processing is already a big task on itself, but this would also require a rewrite of both the Interpreter class and the respective methods. |
+
+| ID                     | DP3 |
+| ---------------------- | ------------------------------------------------------------ |
```

Comment:
Here the chain of responsability design pattern is not really implemented, as the client fowards all instructions directly to the Game, without any processing over it. Furthermore, the Game always processes the Instruction and not passes the responsability along the chain. In the chain of resposability design pattern, the instruction should have been passed on a chain and each member should only know the next one in the chain